lib/ansible/inventory/group.py

```diff
@@ -98,21 +98,34 @@ def _walk_relationship(self, rel):
         '''
         seen = set([])
         unprocessed = set(getattr(self, rel))
+        if include_self:
+            unprocessed.add(self)
+        if preserve_ordering:
+            ordered = [self] if include_self else []
```

Comment:
instead, do

```python
ordered = list(unprocessed)
```

And delay adding `getattr(self, rel)` until after. Maybe a little cleaner...